pkg/config/dynamicloader_test.go

```diff
@@ -0,0 +1,612 @@
+
+	loaderCfg := LoaderConfig{
+		Sources:       nil,
+		TemplatePaths: []string{fileFS},
+	}
+	cmf := generateLoader(t, loaderCfg)
+	cfg := &Config{}
+	err := cmf.ProcessConfigs(cfg, nil)
+	assert.Nil(t, err)
+	assert.Len(t, cfg.Metrics.Configs, 2)
+}
+
+func TestConfigMakerWithExporter(t *testing.T) {
+	configStr := `
+windows_exporter:
+  enabled_collectors: one,two,three
+`
+	tDir := generatePath(t)
+	writeFile(t, tDir, "integrations-1.yml", configStr)
+	fileFS := generateFilePath(tDir)
+
+	loaderCfg := LoaderConfig{
+		Sources:       nil,
+		TemplatePaths: []string{fileFS},
+	}
+	cmf := generateLoader(t, loaderCfg)
+	configs, err := cmf.processIntegrations()
+	assert.NoError(t, err)
+	assert.Len(t, configs, 1)
+	wincfg, _ := configs[0].(v2.UpgradedConfig).LegacyConfig()
+	assert.True(t, wincfg.(*windows_exporter.Config).EnabledCollectors == "one,two,three")
+}
+
+func TestConfigMakerSingletonWithExporter(t *testing.T) {
+	configStr := `
+windows_exporter:
+  enabled_collectors: one,two,three
+  instance: testinstance
+`
+	tDir := generatePath(t)
+	writeFile(t, tDir, "integrations-1.yml", configStr)
+	writeFile(t, tDir, "integrations-2.yml", configStr)
+
+	fileFS := generateFilePath(tDir)
+
+	loaderCfg := LoaderConfig{
+		Sources:       nil,
+		TemplatePaths: []string{fileFS},
+	}
+	cmf := generateLoader(t, loaderCfg)
+	_, err := cmf.processIntegrations()
+	assert.Error(t, err)
+	assert.True(t, strings.Contains(err.Error(), "found multiple instances of singleton"))
+}
+
+func TestConfigMakerWithExporterWithTemplate(t *testing.T) {
+	configStr := `
+windows_exporter:
+  enabled_collectors: {{ (datasource "vars").value }}
+  instance: testinstance
+`
+	tDir := generatePath(t)
+	writeFile(t, tDir, "vars.yaml", "value: banana")
+	writeFile(t, tDir, "integrations-1.yml", configStr)
+	fileFS := generateFilePath(tDir)
+
+	loaderCfg := LoaderConfig{
+		Sources: []Datasource{{
+			Name: "vars",
+			URL:  generateFilePath(filepath.Join(tDir, "vars.yaml")),
+		}},
+		TemplatePaths: []string{fileFS},
+	}
+	cmf := generateLoader(t, loaderCfg)
+	configs, err := cmf.processIntegrations()
+	assert.NoError(t, err)
+	assert.Len(t, configs, 1)
+	wincfg, _ := configs[0].(v2.UpgradedConfig).LegacyConfig()
+	assert.True(t, wincfg.(*windows_exporter.Config).EnabledCollectors == "banana")
+}
+
+func TestConfigMakerWithMultipleExporter(t *testing.T) {
+	configStr := `
+windows_exporter:
+  enabled_collectors: one,two,three
+  instance: testinstance
+node_exporter:
+  autoscrape:
+    enable: false
+`
+	tDir := generatePath(t)
+	writeFile(t, tDir, "integrations-1.yml", configStr)
+	fileFS := generateFilePath(tDir)
+
+	loaderCfg := LoaderConfig{
+		Sources:       nil,
+		TemplatePaths: []string{fileFS},
+	}
+	cmf := generateLoader(t, loaderCfg)
+	configs, err := cmf.processIntegrations()
+	assert.NoError(t, err)
+	assert.Len(t, configs, 2)
+	for _, cfg := range configs {
+		switch v := cfg.(type) {
+		default:
+			fmt.Printf("unexpected type %T", v)
```

Comment:
"Useless comment of the week" candidate here, but did we mean for this to t.Errorf if a different type was encountered?

Reply:
t.Errorf is far better! Thanks